includes/Core/Util/Google_URL_Matcher_Trait.php

```diff
@@ -0,0 +1,87 @@
+<?php
+/**
+ * Trait Google\Site_Kit\Core\Util\Google_URL_Matcher_Trait
+ *
+ * @package   Google\Site_Kit\Core\Util
+ * @copyright 2020 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Util;
+
+/**
+ * Trait for matching URLs and domains for Google Site Verification and Search Console.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+trait Google_URL_Matcher_Trait {
+
+	/**
+	 * Compares two URLs for whether they qualify for a Site Verification or Search Console URL match.
+	 *
+	 * In order for the URLs to be considered a match, they have to be fully equal, except for a potential
+	 * trailing slash in one of them, which will be ignored.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $url     The URL.
+	 * @param string $compare The URL to compare.
+	 * @return bool True if the URLs are considered a match, false otherwise.
+	 */
+	protected function is_url_match( $url, $compare ) {
+		$url     = untrailingslashit( $url );
+		$compare = untrailingslashit( $compare );
+
+		return $url === $compare;
+	}
+
+	/**
+	 * Compares two domains for whether they qualify for a Site Verification or Search Console domain match.
+	 *
+	 * The value to compare may be either a domain or a full URL. If the latter, its scheme and a potential trailing
+	 * slash will be stripped out before the comparison.
+	 *
+	 * In order for the comparison to be considered a match then, the domains have to fully match, except for a
+	 * potential "www." prefix, which will be ignored. If the value to compare is a full URL and includes a path other
+	 * than just a trailing slash, it will not be a match.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $domain  A domain.
+	 * @param string $compare The domain or URL to compare.
+	 * @return bool True if the URLs/domains are considered a match, false otherwise.
+	 */
+	protected function is_domain_match( $domain, $compare ) {
+		$domain  = $this->strip_domain_www( $domain );
```

Comment:
The logic here matches the server and looks good. 

Question (for here and server): could `$domain` ever have a trailing slash here? If so, should we `untrailingslashit` it as well?

Reply:
@adamsilverstein It's always a Site Verification / Search Console domain, so it would never have a trailing slash. I intentionally named that variable `$domain` to clarify that.

Reply:
Given the current usage, no. It's only used for Search Console domain properties after removing the `sc-domain:` prefix (leaving the bare domain) and for `INET_DOMAIN` type sites with Site Verification where the site identifier is a bare domain.

I thought maybe we could call `is_url_match` from `is_domain_match` to do this without extra code but I feel like that might cause problems later if things change.

Anyways, `$domain` should always be given a bare domain, so if it isn't then I would expect the function to not work properly 😄